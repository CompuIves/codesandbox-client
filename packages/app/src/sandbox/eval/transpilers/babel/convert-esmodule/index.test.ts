import { convertEsModule } from '.';

describe('convert-esmodule', () => {
  it('can convert reexports', () => {
    const code = `
      export * from './test/store.js';
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can convert normal exports', () => {
    const code = `
      export { test, test2 } from './test/store.js';
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can convert function exports', () => {
    const code = `
      export function test() {}
      export const test2 = () => {}
      export class Test {}
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can convert default exports', () => {
    const code = `
      export default function test() {}
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can convert class default exports', () => {
    const code = `
      export default class A {}
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can convert weird default exports', () => {
    const code = `
      export default a = 'b';
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can convert default imports', () => {
    const code = `
      import a from './b';
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can convert named imports', () => {
    const code = `
      import {a, b, c} from './b';

      a();
      b();
      c();
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can convert named imports with different scopes', () => {
    const code = `
      import {a} from './b';

      a();

      function test1() {
        a();
      }
      function test2(a) {
        a();

        function test3() {
          a();
        }
      }
    `;

    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can handle as imports', () => {
    const code = `
      import { a as b } from './b';

      const func = b();
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('ignores comments', () => {
    const code = `
      // import { a as b } from './b';
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can handle inline comments', () => {
    const code = `
      import { a as b, /* HELLO WORLD */ c } from './b';

      b();
      c();
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can handle class properties', () => {
    const code = `
      import T from './test';
      class T2 {
        a = () => {
          return 'test'
        }
        b = ''
        c = {}
        static d = ''
      }
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('can handle async code', () => {
    const code = `
      import T from './test';
      async () => {
        const test = await test2()
      }
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('handles export mutations', () => {
    const code = `
      export default function test() {}

      test = 5;
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('handles export mutations with no named function', () => {
    const code = `
      export default function() {}
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('handles export mutations with variables', () => {
    const code = `
    export var to;

    function assign() {
      to = "test"
    }

    function assign2(to) {
      to = "test"
    }
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it("doesn't remove object initializers", () => {
    const code = `
    import { defineHidden, is, createInterpolator, each, getFluidConfig, isAnimatedString, useForceUpdate } from '@react-spring/shared';

    const createHost = (components, {
      a = () => {}
    } = {}) => {
     is()
    };
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it("doesn't set var definitions", () => {
    const code = `
    export var global = typeof window !== 'undefined' ? window : {};
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('handles default as exports', () => {
    const code = `
    export { default as Field } from './Field';
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('handles named exports', () => {
    const code = `
    const a = 'c';
    export { a };
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('handles re-exports in named exports', () => {
    const code = `
    import { a } from './b';
    export { a };
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('handles re-exports in named exports with a alias', () => {
    const code = `
    import { a } from './b';
    const c = 'test';
    export { a as b, c };
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('handles default imports', () => {
    const code = `
    import * as React from 'react';

    console.log(React.Component);
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('handles multiple aliased exports', () => {
    const code = `
    export { _getArrayObserver as getArrayObserver, a as b };
    export { _getMapObserver as getMapObserver, c as d };
    export { _getSetObserver as getSetObserver, e as f };

    f.test();
    d.test();
    b.test();
    `;

    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('converts object shorthands', () => {
    const code = `
    import { templateFactory } from './template-factory.js';

    const short = { templateFactory };
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('hoists imports at bottom', () => {
    const code = `
    const a = PropTypes.a;

    import PropTypes from 'prop-types';
    `;
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('parses and writes chars with linebreaks', () => {
    const code =
      "var WS_CHARS = 'u2000-\\u200a\\u2028\\u2029\\u202f\\u205f\\u3000\\ufeff'";
    expect(convertEsModule(code)).toMatchSnapshot();
  });

  it('has good perf', () => {
    /* eslint-disable */
    const code = require('./big-file');

    const t = Date.now();
    const n = 5;

    for (let i = 0; i < n; i++) {
      convertEsModule(code);
    }
    console.log((Date.now() - t) / n);
    /* eslint-enable */
  });

  it('handles import statement after default export', () => {
    const code = `
    export default function defaultOverscanIndicesGetter(_ref) {
    }

    import { bpfrpt_proptype_OverscanIndicesGetterParams } from './types';
    `;

    convertEsModule(code);
  });
});
