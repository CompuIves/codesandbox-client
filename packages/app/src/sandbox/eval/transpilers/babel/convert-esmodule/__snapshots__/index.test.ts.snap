// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert-esmodule can convert class default exports 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
class A {}
exports.default = A;
"
`;

exports[`convert-esmodule can convert default exports 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
function test() {}
exports.default = test;
"
`;

exports[`convert-esmodule can convert default imports 1`] = `
"\\"use strict\\";
var $csb__b = require(\\"./b\\");
var a = $_csb__interopRequireDefault($csb__b);
function $_csb__interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    default: obj
  };
}
"
`;

exports[`convert-esmodule can convert function exports 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
function test() {}
exports.test = test;
const test2 = () => {};
exports.test2 = test2;
class Test {}
exports.Test = Test;
"
`;

exports[`convert-esmodule can convert named imports 1`] = `
"\\"use strict\\";
var $csb__b = require(\\"./b\\");
$csb__b.a();
$csb__b.b();
$csb__b.c();
"
`;

exports[`convert-esmodule can convert named imports with different scopes 1`] = `
"\\"use strict\\";
var $csb__b = require(\\"./b\\");
$csb__b.a();
function test1() {
  $csb__b.a();
}
function test2(a) {
  a();
  function test3() {
    a();
  }
}
"
`;

exports[`convert-esmodule can convert normal exports 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
var $csb__store = require(\\"./test/store.js\\");
Object.defineProperty(exports, \\"test\\", {
  enumerable: true,
  configurable: true,
  get: function get() {
    return $csb__store.test;
  }
});
Object.defineProperty(exports, \\"test2\\", {
  enumerable: true,
  configurable: true,
  get: function get() {
    return $csb__store.test2;
  }
});
"
`;

exports[`convert-esmodule can convert normal exports 2`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
var $csb__some = require(\\"./some.js\\");
Object.keys($csb__some).forEach(function (key) {
  if (key === \\"default\\" || key === \\"__esModule\\") return;
  if (Object.prototype.hasOwnProperty.call(exports, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    configurable: true,
    get: function get() {
      return $csb__some[key];
    }
  });
});
var $csb__some_ = $_csb__interopRequireDefault(require(\\"./some.js\\"));
Object.defineProperty(exports, \\"some\\", {
  enumerable: true,
  configurable: true,
  get: function get() {
    return $csb__some_.default;
  }
});
function $_csb__interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    default: obj
  };
}
"
`;

exports[`convert-esmodule can convert reexports 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
var $csb__store = require(\\"./test/store.js\\");
Object.keys($csb__store).forEach(function (key) {
  if (key === \\"default\\" || key === \\"__esModule\\") return;
  if (Object.prototype.hasOwnProperty.call(exports, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    configurable: true,
    get: function get() {
      return $csb__store[key];
    }
  });
});
"
`;

exports[`convert-esmodule can convert weird default exports 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
var $csb__default = a = \\"b\\";
exports.default = $csb__default;
"
`;

exports[`convert-esmodule can handle as imports 1`] = `
"\\"use strict\\";
var $csb__b = require(\\"./b\\");
const func = $csb__b.a();
"
`;

exports[`convert-esmodule can handle async code 1`] = `
"\\"use strict\\";
var $csb__test = require(\\"./test\\");
var T = $_csb__interopRequireDefault($csb__test);
(async () => {
  const test = await test2();
});
function $_csb__interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    default: obj
  };
}
"
`;

exports[`convert-esmodule can handle class properties 1`] = `
"\\"use strict\\";
var $csb__test = require(\\"./test\\");
var T = $_csb__interopRequireDefault($csb__test);
class T2 {
  a = () => {
    return \\"test\\";
  }
  b = \\"\\"
  c = {}
  static d = \\"\\"
}
function $_csb__interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    default: obj
  };
}
"
`;

exports[`convert-esmodule can handle inline comments 1`] = `
"\\"use strict\\";
var $csb__b = require(\\"./b\\");
$csb__b.a();
$csb__b.c();
"
`;

exports[`convert-esmodule changes default imports inline 1`] = `
"\\"use strict\\";
var $csb__rgb = require(\\"./rgb\\");
var rgb = $_csb__interopRequireDefault($csb__rgb);
rgb.default.a;
function $_csb__interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    default: obj
  };
}
"
`;

exports[`convert-esmodule converts object shorthands 1`] = `
"\\"use strict\\";
var $csb__templatefactory = require(\\"./template-factory.js\\");
const short = {
  templateFactory: $csb__templatefactory.templateFactory
};
"
`;

exports[`convert-esmodule does empty exports 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
var $csb__column_sorting_draggable = require(\\"./column_sorting_draggable\\");
var $csb__data_grid = require(\\"./data_grid\\");
Object.defineProperty(exports, \\"EuiDataGrid\\", {
  enumerable: true,
  configurable: true,
  get: function get() {
    return $csb__data_grid.EuiDataGrid;
  }
});
var $csb__data_grid_types = require(\\"./data_grid_types\\");
Object.keys($csb__data_grid_types).forEach(function (key) {
  if (key === \\"default\\" || key === \\"__esModule\\") return;
  if (Object.prototype.hasOwnProperty.call(exports, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    configurable: true,
    get: function get() {
      return $csb__data_grid_types[key];
    }
  });
});
"
`;

exports[`convert-esmodule doesn't remove object initializers 1`] = `
"\\"use strict\\";
var $csb__shared = require(\\"@react-spring/shared\\");
const createHost = (components, {a = () => {}} = {}) => {
  $csb__shared.is();
};
"
`;

exports[`convert-esmodule doesn't set var definitions 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
var global = typeof window !== \\"undefined\\" ? window : {};
exports.global = global;
"
`;

exports[`convert-esmodule exports that are not on the root scope are not renamed 1`] = `
"\\"use strict\\";
function a() {
  var exports = \\"blaat\\";
}
"
`;

exports[`convert-esmodule generates parseable var name with @ 1`] = `
"\\"use strict\\";
var $csb__akjaw = require(\\"./a-@kjaw\\");
"
`;

exports[`convert-esmodule handles assignments and exports at the same time 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
const {Ease: C, Linear: Linear, Power0: Power0, Power1: Power1, Power2: Power2, Power3: Power3, Power4: Power4, TweenPlugin: TweenPlugin} = _gsScope;
exports.C = C;
exports.Linear = Linear;
exports.Power0 = Power0;
exports.Power1 = Power1;
exports.Power2 = Power2;
exports.Power3 = Power3;
exports.Power4 = Power4;
exports.TweenPlugin = TweenPlugin;
"
`;

exports[`convert-esmodule handles concurrent import and exports 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"a\\", {
  enumerable: true,
  configurable: true,
  get: function get() {
    return a;
  }
});
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
var $csb__testlibdom = require(\\"test-lib-dom\\");
var $csb__testlibdom_ = require(\\"test-lib-dom\\");
Object.keys($csb__testlibdom_).forEach(function (key) {
  if (key === \\"default\\" || key === \\"__esModule\\") return;
  if (Object.prototype.hasOwnProperty.call(exports, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    configurable: true,
    get: function get() {
      return $csb__testlibdom_[key];
    }
  });
});
var a = () => $csb__testlibdom.a;
"
`;

exports[`convert-esmodule handles default as exports 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
var $csb__Field = $_csb__interopRequireDefault(require(\\"./Field\\"));
Object.defineProperty(exports, \\"Field\\", {
  enumerable: true,
  configurable: true,
  get: function get() {
    return $csb__Field.default;
  }
});
function $_csb__interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    default: obj
  };
}
"
`;

exports[`convert-esmodule handles default imports 1`] = `
"\\"use strict\\";
var $csb__react = require(\\"react\\");
var React = $csb__react;
console.log(React.Component);
"
`;

exports[`convert-esmodule handles export alias 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"default\\", {
  enumerable: true,
  configurable: true,
  get: function get() {
    return Subscription;
  }
});
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
"
`;

exports[`convert-esmodule handles export mutations 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
function test() {}
exports.default = test;
exports.default = test = 5;
"
`;

exports[`convert-esmodule handles export mutations with no named function 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
function $csb__default() {}
exports.default = $csb__default;
"
`;

exports[`convert-esmodule handles export mutations with variables 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
var to;
exports.to = to;
function assign() {
  exports.to = to = \\"test\\";
}
function assign2(to) {
  to = \\"test\\";
}
"
`;

exports[`convert-esmodule handles export object 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"c\\", {
  enumerable: true,
  configurable: true,
  get: function get() {
    return c;
  }
});
Object.defineProperty(exports, \\"b\\", {
  enumerable: true,
  configurable: true,
  get: function get() {
    return b;
  }
});
Object.defineProperty(exports, \\"a\\", {
  enumerable: true,
  configurable: true,
  get: function get() {
    return a;
  }
});
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
"
`;

exports[`convert-esmodule handles import statement after default export 1`] = `
"\\"use strict\\";
var $csb__types = require(\\"./types\\");
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
function defaultOverscanIndicesGetter(_ref) {}
exports.default = defaultOverscanIndicesGetter;
"
`;

exports[`convert-esmodule handles multiple aliased exports 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"f\\", {
  enumerable: true,
  configurable: true,
  get: function get() {
    return e;
  }
});
Object.defineProperty(exports, \\"getSetObserver\\", {
  enumerable: true,
  configurable: true,
  get: function get() {
    return _getSetObserver;
  }
});
Object.defineProperty(exports, \\"d\\", {
  enumerable: true,
  configurable: true,
  get: function get() {
    return c;
  }
});
Object.defineProperty(exports, \\"getMapObserver\\", {
  enumerable: true,
  configurable: true,
  get: function get() {
    return _getMapObserver;
  }
});
Object.defineProperty(exports, \\"b\\", {
  enumerable: true,
  configurable: true,
  get: function get() {
    return a;
  }
});
Object.defineProperty(exports, \\"getArrayObserver\\", {
  enumerable: true,
  configurable: true,
  get: function get() {
    return _getArrayObserver;
  }
});
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
f.test();
d.test();
b.test();
"
`;

exports[`convert-esmodule handles multiple var exports 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
const a = 5, b = 6;
exports.a = a;
exports.b = b;
"
`;

exports[`convert-esmodule handles named exports 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"a\\", {
  enumerable: true,
  configurable: true,
  get: function get() {
    return a;
  }
});
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
const a = \\"c\\";
"
`;

exports[`convert-esmodule handles re-exports in named exports 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"a\\", {
  enumerable: true,
  configurable: true,
  get: function get() {
    return $csb__b.a;
  }
});
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
var $csb__b = require(\\"./b\\");
"
`;

exports[`convert-esmodule handles re-exports in named exports with a alias 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"c\\", {
  enumerable: true,
  configurable: true,
  get: function get() {
    return c;
  }
});
Object.defineProperty(exports, \\"b\\", {
  enumerable: true,
  configurable: true,
  get: function get() {
    return $csb__b.a;
  }
});
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
var $csb__b = require(\\"./b\\");
const c = \\"test\\";
"
`;

exports[`convert-esmodule hoists imports at bottom 1`] = `
"\\"use strict\\";
var $csb__proptypes = require(\\"prop-types\\");
var PropTypes = $_csb__interopRequireDefault($csb__proptypes);
const a = PropTypes.default.a;
function $_csb__interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : {
    default: obj
  };
}
"
`;

exports[`convert-esmodule ignores comments 1`] = `
"\\"use strict\\";
"
`;

exports[`convert-esmodule keeps import order 1`] = `
"\\"use strict\\";
var $csb__1 = require(\\"1\\");
var $csb__2 = require(\\"2\\");
"
`;

exports[`convert-esmodule parses and writes chars with linebreaks 1`] = `
"\\"use strict\\";
var WS_CHARS = \\"u2000-     　﻿\\";
"
`;

exports[`convert-esmodule renames exports that are already defined, even in block scope 1`] = `
"\\"use strict\\";
var __$csb_exports = \\"testtest\\";
function a() {
  __$csb_exports = \\"blaat\\";
}
"
`;

exports[`convert-esmodule works with variables that are named exports 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
var __$csb_exports = [eventedState, eventedShowHideState];
__$csb_exports.push(\\"test\\");
var $csb__default = __$csb_exports;
exports.default = $csb__default;
"
`;
